File: designMode/singletion-mode/singletion-mode.js
// class Singletion {
//     constructor(name) {
//         this.name = name;
//     }
// }

// function createAdmin(name) {
//     if (!Singletion.instance) {
//         Singletion.instance = new Singletion(name);
//     }
//     return Singletion.instance; 
// }

// let admin1 = createAdmin("js");
// let admin2 = createAdmin("node");

// console.log(admin1 === admin2);
// console.log("admin1 : ", admin1.name);
// console.log("admin2 : ", admin2.name);


// class Singletion {
//     constructor(name) {
//         if (!Singletion.instance) {
//             this.name = name;
//             Singletion.instance = this;
//         }
//         return Singletion.instance;
//     }

//     printName() {
//         console.log(this.name);
//     }
// }

// let admin1 = new Singletion("js");
// let admin2 = new Singletion("node");

// console.log(admin1 === admin2);
// console.log("admin1 : ", admin1.name);
// console.log("admin2 : ", admin2.name);

// admin1.printName();
// admin2.printName();

// let admin = {
//     _name : 'js',
//     setName : (name) => {
//         this._name = name;
//     },
//     getName : () => {
//         return this._name;
//     }
// };

let adminName = (function Singletion(name) {
    return () => {
        return name;
    };
})("javascript");

console.log(adminName());

File: javascript/designMode/singletion-mode/singletion-mode.c
#include <stdio.h>

// static char admin_name[20] = "admin";

// static void set_admin_name(char *name) {
//     sprintf(admin_name, "%s", name);
// }

// static char *get_admin_name(void) {
//     return admin_name;
// }

// int main(int args, char** argv) {
//     set_admin_name("javascript");
//     printf("%s\n", get_admin_name());
//     return 0;
// }

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

static char *create_admin(char *name) {
    static bool is_created = false;
    static char *admin = NULL;
    if (!is_created) {
        admin = (char *)malloc(20);
        sprintf(admin, "%s", name);
        is_created = true;
    }
    return admin;
}

int main(int argc, char** argv) {
    char *admin1 = create_admin("javascript");
    char *admin2 = create_admin("c language");
    printf("%s\n", (admin1 == admin2) ? "true" : "false");
    printf("admin1 %s\n", admin1);
    printf("admin2 %s\n", admin2);
    return 0;
}
